### ISSUE TITLE
About model configuration.

### ISSUE THREAD

**Issue by dxu23nc** [ray-project/ray#7644]
I noticed that there are two model configuration parameters: fcnet_hiddens in model_config and hiddens in the configuration of dqn.py. 
I try to print the whole model (model.base_model.summary()), it shows that the "hiddens" do not show up. Why did that happen? I know this "hiddens" is used in function build_q_model dqn.py. The code is as follows:
`
import ray
from ray import tune
import functools
import pickle
import threading

from ray.rllib.agents import dqn
ray.init()
config= dqn.DEFAULT_CONFIG.copy()
config["num_workers"] = 1
config["hiddens"] = [512,128] 
trainer = dqn.DQNTrainer(config=config,env="AirRaid-ram-v0")
model = trainer.get_policy().model
model.base_model.summary()
model.q_value_head.summary()
` 

And the log is:
Model: "model"
__________________________________________________________________________________________________
Layer (type)                    Output Shape         Param      Connected to

observations (InputLayer)       [(None, 128)]        0
__________________________________________________________________________________________________
fc_1 (Dense)                    (None, 256)          33024       observations[0][0]
__________________________________________________________________________________________________
fc_out (Dense)                  (None, 256)          65792       fc_1[0][0]
__________________________________________________________________________________________________
value_out (Dense)               (None, 1)            257         fc_1[0][0]

Total params: 99,073
Trainable params: 99,073
Non-trainable params: 0
__________________________________________________________________________________________________
Model: "model_1"
_________________________________________________________________
Layer (type)                 Output Shape              Param 
model_out (InputLayer)       [(None, 256)]             0
_________________________________________________________________
lambda (Lambda)              [(None, 6), (None, 6, 1), 198022
Total params: 198,022
Trainable params: 198,022
Non-trainable params: 0
_________________________________________________________________

**Comment by rfali** [comment]
@ericl @richardliaw It seems that the above issue has gone away with recent ray versions, not sure what resolved it. 

ray version 1.3.0, 
tensorflow 2.34.1

If I run the above code
```python

import ray
from ray.rllib.agents import dqn

ray.init()
config= dqn.DEFAULT_CONFIG.copy()
config["num_workers"] = 1
config["hiddens"] = [512,128]
trainer = dqn.DQNTrainer(config=config,env="AirRaid-ram-v0")
model = trainer.get_policy().model
model.base_model.summary()
model.q_value_head.summary()
```
I get the following output
```
Model: "model"
__________________________________________________________________________________________________
Layer (type)                    Output Shape         Param #     Connected to                     
==================================================================================================
observations (InputLayer)       [(None, 128)]        0                                            
__________________________________________________________________________________________________
fc_1 (Dense)                    (None, 256)          33024       observations[0][0]               
__________________________________________________________________________________________________
fc_out (Dense)                  (None, 256)          65792       fc_1[0][0]                       
__________________________________________________________________________________________________
value_out (Dense)               (None, 1)            257         fc_1[0][0]                       
==================================================================================================
Total params: 99,073
Trainable params: 99,073
Non-trainable params: 0
__________________________________________________________________________________________________
Model: "model_1"
__________________________________________________________________________________________________
Layer (type)                    Output Shape         Param #     Connected to                     
==================================================================================================
model_out (InputLayer)          [(None, 256)]        0                                            
__________________________________________________________________________________________________
hidden_0 (Dense)                (None, 512)          131584      model_out[0][0]                  
__________________________________________________________________________________________________
hidden_1 (Dense)                (None, 128)          65664       hidden_0[0][0]                   
__________________________________________________________________________________________________
dense (Dense)                   (None, 6)            774         hidden_1[0][0]                   
__________________________________________________________________________________________________
tf_op_layer_default_policy/ones [(2,)]               0           dense[0][0]                      
__________________________________________________________________________________________________
tf_op_layer_default_policy/ones [(2,)]               0           dense[0][0]                      
__________________________________________________________________________________________________
tf_op_layer_default_policy/ones [(None, 6)]          0           tf_op_layer_default_policy/ones_l
__________________________________________________________________________________________________
tf_op_layer_default_policy/ones [(None, 6)]          0           tf_op_layer_default_policy/ones_l
__________________________________________________________________________________________________
tf_op_layer_default_policy/Expa [(None, 6, 1)]       0           tf_op_layer_default_policy/ones_l
__________________________________________________________________________________________________
tf_op_layer_default_policy/Expa [(None, 6, 1)]       0           tf_op_layer_default_policy/ones_l
==================================================================================================
Total params: 198,022
Trainable params: 198,022
Non-trainable params: 0
```
which seems accurate to me. The above output corresponds to
```
[base_model] obs_space_size ->  2 dense layers of 256 each (last being called fc_out)  
[q_head] fc_out -> 2 q_hidden layers [512, 128] -> Q_values of action_space_size  

i.e.
[128 observations] -> [256] -> [256] -> [512] -> [128] -> [6 actions]
```
This is because the default value of `fcnet_hiddens`  is [256,256]. The last few layers in the q_head correspond to the layers if distributional Q-learning (C-51) is used.

The `q_hiddens` or basically `config["hiddens"]` (this inconsistency should be addressed) is the number of dense layers after the Advantage-Value split, as documented in the [DistributionalQTFModel](https://github.com/ray-project/ray/blob/340fbf53c0e885c8406d20679e6917d6f6120f0d/rllib/agents/dqn/distributional_q_tf_model.py#L49).

If you set Dueling to True

```python

import ray
from ray.rllib.agents import dqn

ray.init()
config= dqn.DEFAULT_CONFIG.copy()
config["num_workers"] = 1
config["hiddens"] = [512,128]
config["dueling"] = True

trainer = dqn.DQNTrainer(config=config,env="AirRaid-ram-v0")
model = trainer.get_policy().model
model.base_model.summary()
model.q_value_head.summary()
model.state_value_head.summary()
```
we get
```
Model: "model"
__________________________________________________________________________________________________
Layer (type)                    Output Shape         Param #     Connected to                     
==================================================================================================
observations (InputLayer)       [(None, 128)]        0                                            
__________________________________________________________________________________________________
fc_1 (Dense)                    (None, 256)          33024       observations[0][0]               
__________________________________________________________________________________________________
fc_out (Dense)                  (None, 256)          65792       fc_1[0][0]                       
__________________________________________________________________________________________________
value_out (Dense)               (None, 1)            257         fc_1[0][0]                       
==================================================================================================
Total params: 99,073
Trainable params: 99,073
Non-trainable params: 0
__________________________________________________________________________________________________
Model: "model_1"
__________________________________________________________________________________________________
Layer (type)                    Output Shape         Param #     Connected to                     
==================================================================================================
model_out (InputLayer)          [(None, 256)]        0                                            
__________________________________________________________________________________________________
hidden_0 (Dense)                (None, 512)          131584      model_out[0][0]                  
__________________________________________________________________________________________________
hidden_1 (Dense)                (None, 128)          65664       hidden_0[0][0]                   
__________________________________________________________________________________________________
dense (Dense)                   (None, 6)            774         hidden_1[0][0]                   
__________________________________________________________________________________________________
tf_op_layer_default_policy/ones [(2,)]               0           dense[0][0]                      
__________________________________________________________________________________________________
tf_op_layer_default_policy/ones [(2,)]               0           dense[0][0]                      
__________________________________________________________________________________________________
tf_op_layer_default_policy/ones [(None, 6)]          0           tf_op_layer_default_policy/ones_l
__________________________________________________________________________________________________
tf_op_layer_default_policy/ones [(None, 6)]          0           tf_op_layer_default_policy/ones_l
__________________________________________________________________________________________________
tf_op_layer_default_policy/Expa [(None, 6, 1)]       0           tf_op_layer_default_policy/ones_l
__________________________________________________________________________________________________
tf_op_layer_default_policy/Expa [(None, 6, 1)]       0           tf_op_layer_default_policy/ones_l
==================================================================================================
Total params: 198,022
Trainable params: 198,022
Non-trainable params: 0
__________________________________________________________________________________________________
Model: "model_2"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
=================================================================
model_out (InputLayer)       [(None, 256)]             0         
_________________________________________________________________
dense_1 (Dense)              (None, 512)               131584    
_________________________________________________________________
dense_2 (Dense)              (None, 128)               65664     
_________________________________________________________________
dense_3 (Dense)              (None, 1)                 129       
=================================================================
Total params: 197,377
Trainable params: 197,377
Non-trainable params: 0
```
which corresponds to
```
[128 observations] -> [256] -> [256] -> [512] -> [128] -> [6 Q values for the 6 possible actions]
                                                         -> [512] -> [128] -> [1 value]            
```
The last `[256]` is the common layer, and splits into two branches, one each for Advantage and Value streams, as described in the Dueling DQN paper.

**Comment by rfali** [comment]
However, I have stumbled into a corner case, which could be a lack of understanding on my part. If you set `hiddens` to be an empty list, and set `dueling` True, then the `V(s)` is used from the `state_value_head` [here](https://github.com/ray-project/ray/blob/340fbf53c0e885c8406d20679e6917d6f6120f0d/rllib/agents/dqn/distributional_q_tf_model.py#L167). This makes for an interesting case since this value is not connected to the last shared dense layer, but actually the Q_head, which would be wrong. Here is a reproduction code.

```python

# CartPole with q-hiddens ON and OFF

import ray
from ray.rllib.agents import dqn

ray.init()
config= dqn.DEFAULT_CONFIG.copy()
config["num_workers"] = 1
config["model"]["fcnet_hiddens"] = [64, 64]
config["hiddens"] = []  #[128, 128]
config["dueling"] = True

trainer = dqn.DQNTrainer(config=config,env="CartPole-v0")
model = trainer.get_policy().model
model.base_model.summary()
model.q_value_head.summary()
model.state_value_head.summary()

# 4a. Output with q-hiddens ON [128, 128]
"""
Model: "model"
__________________________________________________________________________________________________
Layer (type)                    Output Shape         Param #     Connected to                     
==================================================================================================
observations (InputLayer)       [(None, 4)]          0                                            
__________________________________________________________________________________________________
fc_1 (Dense)                    (None, 64)           320         observations[0][0]               
__________________________________________________________________________________________________
fc_out (Dense)                  (None, 64)           4160        fc_1[0][0]                       
__________________________________________________________________________________________________
value_out (Dense)               (None, 1)            65          fc_1[0][0]                       
==================================================================================================
Total params: 4,545
Trainable params: 4,545
Non-trainable params: 0
__________________________________________________________________________________________________
Model: "model_1"
__________________________________________________________________________________________________
Layer (type)                    Output Shape         Param #     Connected to                     
==================================================================================================
model_out (InputLayer)          [(None, 64)]         0                                            
__________________________________________________________________________________________________
hidden_0 (Dense)                (None, 128)          8320        model_out[0][0]                  
__________________________________________________________________________________________________
hidden_1 (Dense)                (None, 128)          16512       hidden_0[0][0]                   
__________________________________________________________________________________________________
dense (Dense)                   (None, 2)            258         hidden_1[0][0]                   
__________________________________________________________________________________________________
tf_op_layer_default_policy/ones [(2,)]               0           dense[0][0]                      
__________________________________________________________________________________________________
tf_op_layer_default_policy/ones [(2,)]               0           dense[0][0]                      
__________________________________________________________________________________________________
tf_op_layer_default_policy/ones [(None, 2)]          0           tf_op_layer_default_policy/ones_l
__________________________________________________________________________________________________
tf_op_layer_default_policy/ones [(None, 2)]          0           tf_op_layer_default_policy/ones_l
__________________________________________________________________________________________________
tf_op_layer_default_policy/Expa [(None, 2, 1)]       0           tf_op_layer_default_policy/ones_l
__________________________________________________________________________________________________
tf_op_layer_default_policy/Expa [(None, 2, 1)]       0           tf_op_layer_default_policy/ones_l
==================================================================================================
Total params: 25,090
Trainable params: 25,090
Non-trainable params: 0
__________________________________________________________________________________________________
Model: "model_2"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
=================================================================
model_out (InputLayer)       [(None, 64)]              0         
_________________________________________________________________
dense_1 (Dense)              (None, 128)               8320      
_________________________________________________________________
dense_2 (Dense)              (None, 128)               16512     
_________________________________________________________________
dense_3 (Dense)              (None, 1)                 129       
=================================================================
Total params: 24,961
Trainable params: 24,961
Non-trainable params: 0
"""

# 4b. Output with q-hiddens OFF []

"""
Model: "model"
__________________________________________________________________________________________________
Layer (type)                    Output Shape         Param #     Connected to                     
==================================================================================================
observations (InputLayer)       [(None, 4)]          0                                            
__________________________________________________________________________________________________
fc_1 (Dense)                    (None, 64)           320         observations[0][0]               
__________________________________________________________________________________________________
fc_2 (Dense)                    (None, 64)           4160        fc_1[0][0]                       
__________________________________________________________________________________________________
fc_out (Dense)                  (None, 2)            130         fc_2[0][0]                       
__________________________________________________________________________________________________
value_out (Dense)               (None, 1)            65          fc_2[0][0]                       
==================================================================================================
Total params: 4,675
Trainable params: 4,675
Non-trainable params: 0
__________________________________________________________________________________________________
Model: "model_1"
__________________________________________________________________________________________________
Layer (type)                    Output Shape         Param #     Connected to                     
==================================================================================================
model_out (InputLayer)          [(None, 2)]          0                                            
__________________________________________________________________________________________________
tf_op_layer_default_policy/ones [(2,)]               0           model_out[0][0]                  
__________________________________________________________________________________________________
tf_op_layer_default_policy/ones [(2,)]               0           model_out[0][0]                  
__________________________________________________________________________________________________
tf_op_layer_default_policy/ones [(None, 2)]          0           tf_op_layer_default_policy/ones_l
__________________________________________________________________________________________________
tf_op_layer_default_policy/ones [(None, 2)]          0           tf_op_layer_default_policy/ones_l
__________________________________________________________________________________________________
tf_op_layer_default_policy/Expa [(None, 2, 1)]       0           tf_op_layer_default_policy/ones_l
__________________________________________________________________________________________________
tf_op_layer_default_policy/Expa [(None, 2, 1)]       0           tf_op_layer_default_policy/ones_l
==================================================================================================
Total params: 0
Trainable params: 0
Non-trainable params: 0
__________________________________________________________________________________________________
Model: "model_2"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
=================================================================
model_out (InputLayer)       [(None, 2)]               0         
_________________________________________________________________
dense (Dense)                (None, 1)                 3         
=================================================================
Total params: 3
Trainable params: 3
Non-trainable params: 0
_________________________________________________________________
```

The part where the `state_value_head` is connected to the `fc_out` layer of size [2] and 0 parameters is what is bothering me. 

Of course this won't happen with the default value of `hiddens` set already to 256 which enforces a dense layer after the last fcnet_hidden layer, a detail which is obscured for a beginner. 

If the `hiddens` is only for the dueling architecture, it could be off by default.

**Comment by rfali** [comment]
Can this be related (although above question is not about VisionNets)?
PR#[18306](https://github.com/ray-project/ray/pull/18306)
>built-in VisionNets: The value branch - in case we have a shared one - is attached to the action-logits output instead of the feature output (one layer before the action logits output).

**Comment by vedhasam** [comment]
There definitely appears to be some design inconsistency in how the models are ultimately created, when `config["hiddens"]=[]` or when `config["hiddens"]=<something non-empty>`.

If:
```
observations (InputLayer) =[(None, 16)]
config["model"]["fcnet_hiddens"] = [20, 17, 10]
num_actions=3
```
`config["hiddens"]=[]` results in
```
base_model: [Nx16]->[Nx20]->[Nx17]->[Nx10]-> (which splits into [Nx1] and [Nx3])
q_value_head: [Nx3] -> (splits into two of [Nx3x1] )
state_value_head: [Nx3] -> [Nx1]
```
whereas `config["hiddens"]=[256]` results in
```
base_model: [Nx16]->[Nx20]->[Nx17] -> (splits into [Nx1] and into [Nx10]) 
q_value_head: [Nx10]->[Nx256]->(splits into two of [Nx3x1] )
state_value_head: [Nx10]->[Nx256]->[Nx1]
```

With everything, especially the`config["model"]["fcnet_hiddens"]` identical, why should `state_value_head` input size be [Nx3[ in one while [Nx10] in another?
From the consistency point of view, for `state_value_head`
EITHER 
A. `[Nx3]->[Nx1]` (with `config["hiddens"]=[]`) implies that `[Nx3]->[Nx256]->[Nx1]` (with `config["hiddens"]=[256]`)
OR
B. `[Nx10]->[Nx256]->[Nx1]` (with `config["hiddens"]=[256]`) implies that `[Nx10]->[Nx1]` (with `config["hiddens"]=[]`) 

Similar arguments could be made for `q_value_head` for `base_model` in terms of the size of the input they use and produce respectively. 

**Most critically**, 
the `value_head` of `base_model` (with `config["hiddens"]=[256]`) uses the 10-sized fc_net_hiddens[-1], while in the other `base_model` bypassees it completely to compute `value_head`.

**Comment by vedhasam** [comment]
Here's how the two models (`hiddens=[]` v/s `hiddens=[256]`) look like pictorially, with all other hyperparameters identical otherwise.
![image](https://user-images.githubusercontent.com/82571206/211961734-a2d31fa9-e7e5-416f-9cfe-19049bf0ca57.png)

Please correct me if I am wrong, but if one was to assume the 3 constituent models obtained with `hiddens=[256]` to be the right implementation, then with `hiddens=[]`, one would only expect that only the dense layer with 256 nodes gets deleted, with no change in the interface between the 3 constituent models. If you see the interfacing size (output of base_model = input of state_value_head = input of q_value_head=) changes from 10 to 3. Wouldn't it be easier if fcnet_hiddens were to dedicatedly compute `value_out` and `advantages/q` consistently? On the other hand, if we say that fc_out computes features and not advantages or q, wouldn't it be easier if the last layer of fcnet_hiddens (of size 10) was to dedicatedly compute the features consistently in both case (hiddens =[256] v/s []), for the consistency sake?

**Comment by karstenddwx** [comment]
Hi,
 any comments to what vedhasam described?

**Comment by cszhu** [comment]
Hello! This P3 issue has seen no activity in at least a year. It will be closed in 2 weeks as part of ongoing cleanup efforts.

Please remove the `pending-cleanup` label if you believe this issue should remain open.

Thanks for contributing to Ray!
